stan/math/rev/fun/softmax.hpp

```diff
@@ -24,19 +24,20 @@ namespace math {
  * @return Softmax of the input.
  * @throw std::domain_error If the input vector is size 0.
  */
-
-inline Eigen::Matrix<var, Eigen::Dynamic, 1> softmax(
-    const Eigen::Matrix<var, Eigen::Dynamic, 1>& alpha) {
-  check_nonzero_size("softmax", "alpha", alpha);
-
-  arena_matrix<Eigen::VectorXd> res_val = softmax(value_of(alpha));
-  arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, 1>> res = res_val;
-  arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, 1>> alpha_arena = alpha;
-
-  reverse_pass_callback([=]() mutable {
+template <typename Mat, require_rev_matrix_t<Mat>* = nullptr>
+inline plain_type_t<Mat> softmax(const Mat& alpha) {
+  using mat_plain = plain_type_t<Mat>;
+  if (alpha.size() == 0) {
+    return alpha;
+  }
+  arena_t<mat_plain> alpha_arena = alpha;
+  arena_t<Eigen::VectorXd> res_val = softmax(value_of(alpha_arena));
```

Comment:
`Eigen::VectorXd` -> `mat_plain`

Reply:
This is a value, `mat_plain` contains vars ir is var_value.